packages/viz/src/lib/layer/Layer.ts

```diff
@@ -0,0 +1,273 @@
+import { Deck } from '@deck.gl/core';
+import { MVTLayer } from '@deck.gl/geo-layers';
+import { Source } from '../sources/Source';
+import { CARTOSource, DOSource } from '../sources';
+import { DOLayer } from '../deck/DOLayer';
+import { getStyles, StyleProperties, Style } from '../style';
+import { PopupElement } from '../popups/Popup';
+import { StyledLayer } from '../style/layer-style';
+import { CartoLayerError, layerErrorTypes } from '../errors/layer-error';
+import {
+  LayerInteractivity,
+  EventType,
+  InteractionHandler
+} from './LayerInteractivity';
+import { LayerOptions } from './LayerOptions';
+
+export class Layer implements StyledLayer {
+  private _source: Source;
+  private _style: Style;
+  private _options: LayerOptions;
+
+  // Deck.gl Map instance
+  private _deckInstance: Deck | undefined;
+
+  // Instance to the DeckLayer of the instance
+  // It cannot be a reference to (import { Layer } from '@deck.gl/core') because
+  // the typing of getPickinfo method is different from TileLayer and Layer are
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private _deckLayer?: any;
+
+  private _interacitivity: LayerInteractivity;
+
+  constructor(
+    source: string | Source,
+    style: Style | StyleProperties = {},
+    options?: Partial<LayerOptions>
+  ) {
+    this._source = buildSource(source);
+    this._style = buildStyle(style);
+
+    let hoverStyle;
+
+    if (options && options.hoverStyle) {
+      hoverStyle = buildStyle(options.hoverStyle);
+    }
+
+    let clickStyle;
+
+    if (options && options.clickStyle) {
+      clickStyle = buildStyle(options.clickStyle);
+    }
+
+    this._interacitivity = new LayerInteractivity(
+      this,
+      this.getStyle.bind(this),
+      this.setStyle.bind(this),
+      hoverStyle,
+      clickStyle
+    );
```

Comment:
How about a small refactor, having a helper method to do `this._interactivity = buildInteractivity(options)`?